### ISSUE TITLE
[Feature]: Implement RSLAD-style robust distillation in compression phase

### ISSUE THREAD

**Issue by Adit-Jain-srm** [Adit-Jain-srm/NightmareNet#47]
## Problem Statement

The compression phase (`CompressionPhase` in `nightmarenet/training/phases.py`) applies magnitude pruning but does NOT perform the "robust distillation" step described in the README:

> **Compress** - Preserve robustness, shed parameters. Adversarial robust distillation (RSLAD-style) + magnitude pruning.

The actual implementation only does pruning + optional fine-tuning:

```python
class CompressionPhase:
    def run(self, dataloader, optimizer):
        # 1. Apply magnitude pruning
        pruner = MagnitudePruner(pruning_ratio=self.pruning_ratio)
        stats = pruner.apply(self.model)
        # 2. Optional fine-tune after prune
        if self.finetune_after_prune:
            # ... standard fine-tuning loop ...
```

There is no knowledge distillation from the pre-pruned model to the pruned model. The README's claim of "RSLAD-style adversarial robust distillation" is unimplemented. This is significant because:

- Standard pruning followed by fine-tuning does NOT preserve adversarial robustness (known finding from Ye et al. 2019)
- The distillation step is what makes the compression phase part of the robustness mechanism (not just a size reduction)
- The paper draft references this but the code doesn't implement it

**Relevant files:**
- `nightmarenet/training/phases.py` (CompressionPhase.run)
- `nightmarenet/compression/pruning.py` (MagnitudePruner only)

## Proposed Solution

Add a distillation step to CompressionPhase:

1. Before pruning, snapshot the current model as the "teacher"
2. After pruning, run a distillation loop where the pruned "student" learns from the teacher's logits
3. Use KL divergence loss between teacher and student outputs (following RSLAD: use adversarial examples as distillation data, not clean data)
4. Config: `compression.distillation: true`, `compression.distillation_temperature: 4.0`, `compression.distillation_alpha: 0.7`

**Files affected:**
- `nightmarenet/compression/pruning.py` or new `nightmarenet/compression/distillation.py`
- `nightmarenet/training/phases.py` (CompressionPhase)
- `configs/default.yaml` (new config keys)

**Dependencies:** None new (KL divergence is in torch.nn.functional).

## Acceptance Criteria

- [ ] CompressionPhase performs KL distillation from pre-pruned teacher to pruned student
- [ ] Distillation uses adversarially-perturbed inputs (not clean data) per RSLAD
- [ ] Config keys added: `compression.distillation`, `compression.distillation_temperature`, `compression.distillation_alpha`
- [ ] Distillation is optional (backward compatible when disabled)
- [ ] Test: verify distillation reduces KL divergence between teacher and student outputs
- [ ] Documentation: update README compress phase description if behavior changes

## Scope / Difficulty Estimate

Medium-high. ~4-5 hours. Requires understanding the distillation literature and implementing a training loop within the compression phase.

**Comment by RitikaSen27** [comment]
hi @Adit-Jain-srm can i work on this?
Could you please assign it to me?

**Comment by Adit-Jain-srm** [comment]
Sure, assigned @RitikaSen27

**Comment by Adit-Jain-srm** [comment]
@RitikaSen27 It has been more than a week since assignment without PR or progress update, Please confirm your willingness to continue to work on this issue to prevent un-assignment.

**Comment by RitikaSen27** [comment]
Yes @Adit-Jain-srm I am still working on it 
I will create a PR for it by tonight for review.
Sorry for the delay

**Comment by RitikaSen27** [comment]
Hey @Adit-Jain-srm , I've worked on this and opened a PR to fix it.

Here's what was done:
- Created `nightmarenet/compression/distillation.py` implementing RSLAD-style distillation — the pre-pruned model is snapshotted as a frozen teacher, and the pruned student learns from its logits via KL divergence loss
- Distillation runs on adversarially-perturbed inputs (FGSM) per RSLAD, not clean data
- Wired the distillation step into `CompressionPhase.run` in `phases.py` — teacher snapshot happens before pruning, distillation runs after
- Added config keys to `configs/default.yaml`: `distillation`, `distillation_temperature` (default 4.0), `distillation_alpha` (default 0.7), `distillation_epochs`
- Feature is opt-in and backward compatible — existing behaviour is preserved when `distillation: false`

All 492 existing tests pass. Please review the PR and let me know if any changes are needed!
